### DISCUSSION TITLE
NoReferencedTableError on ForeignKey

### DISCUSSION THREAD

**Discussion by MijikHna** [sqlalchemy/sqlalchemy#8236] [Usage Questions]
I am using SqlAlchemy ORM with postgres.
I have some migrations written with alembic. But If I want to save my orm I get following error `sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'profiles.user_id' could not find table 'users' with which to generate a foreign key to target column 'id'`
Here is the related migration:
```
op.create_table(
        'profiles',
        sa.Column('id', sa.Integer, primary_key=True),
        sa.Column(
            'theme',
            sa.Text,
            nullable=False,
            server_default='light'
        ),
        sa.Column('image', sa.Text, nullable=True),
        sa.Column(
            'created_at',
            sa.TIMESTAMP(timezone=True),
            server_default=sa.func.now(),
            nullable=False,
        ),
        sa.Column(
            'updated_at',
            sa.TIMESTAMP(timezone=True),
            server_default=sa.func.now(),
            nullable=False,
        ),
        sa.Column(
            'user_id',
            sa.Integer,
            sa.ForeignKey("users.id"),
            nullable=False
        ),
    )
```
here is my ORM model
```
class ProfileDao(Base):
    __tablename__ = 'profiles'
    __table_args__ = {'schema': 'public'}

    id: Column = Column(Integer, primary_key=True)
    theme: Column = Column(Text, nullable=False)
    image: Column = Column(Text, nullable=False)
    created_at: Column = Column(
        TIMESTAMP(timezone=True),
        nullable=False,
        server_default=func.now()
    )
    updated_at: Column = Column(
        TIMESTAMP(timezone=True),
        nullable=False,
        server_default=func.now()
    )
    user_id: Column = Column(
        Integer,
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False
    )

    user: relationship('UserDao', back_populates='profile')
```
and here how I try to write the the model to db
```
def create_profile(
        self,
        db_session: Session,
        profile: ProfileDto
    ) -> ProfileDao:
        profile_db: ProfileDao = ProfileDao(
            theme=profile.theme,
            image=profile.image,
            user_id=profile.user_id,
        )

        db_session.add(profile_db)
        db_session.flush()

        return profile_db
```
I get this error on `db_session.add(profile_db)`

here is also what I have in postgres
```
                                       Table "public.profiles"
   Column   |           Type           | Collation | Nullable |               Default                
------------+--------------------------+-----------+----------+--------------------------------------
 id         | integer                  |           | not null | nextval('profiles_id_seq'::regclass)
 theme      | text                     |           | not null | 'light'::text
 image      | text                     |           |          | 
 created_at | timestamp with time zone |           | not null | now()
 updated_at | timestamp with time zone |           | not null | now()
 user_id    | integer                  |           | not null | 
Indexes:
    "profiles_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "profiles_user_id_fkey" FOREIGN KEY (user_id) REFERENCES users(id)
Triggers:
    update_user_modtime BEFORE UPDATE ON profiles FOR EACH ROW EXECUTE FUNCTION update_updated_at_column()
````

I was able to overcome this error message by deleting the `ForeignKey()` from `user_id` attribute. But I think it wasn't the idea to remove `ForeignKey()` if the foreign key constraint is already written to db.

**Comment by zzzeek** [comment]
if you make an ORM model that contains ForeignKey() to some table, for the ORM model to be functional, there needs to be an ORM model and/or Table object that this ForeignKey can resolve towards.  In this case, a table called "user".  See for eaxmple the many to one relationship at https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-one.  there's the class, then the ForeignKey which has the thing it refers towards.  need both to be present for an ORM model.

**Reply by MijikHna** [reply]
I have also defined this second table **users** <https://github.com/MijikHna/PLAYAREA3/blob/main/fastapi/src-backend/api/db/models/user_dao.py> and also have migrations for this table <https://github.com/MijikHna/PLAYAREA3/blob/main/fastapi/src-backend/api/db/migrations/versions/309b232de140_create_user_table.py>. The code is not up-to-date. But I guess nothing accept that I have added `relationship()` to orm definition has changed in **users** definition.
